siphon/http_util.py

```diff
@@ -0,0 +1,476 @@
+
+    See Also
+    --------
+    http_get
+    '''
+
+    return BytesIO(create_http_session().get(url, **kwargs).content)
+
+
+def parse_iso_date(s):
+    r'''Parse a string containing an ISO-8601 formatted date
+
+    Parameters
+    ----------
+    s : string
+        The string to be parsed
+
+    Returns
+    -------
+    dt : ``datetime.datetime`` instance
+        The results of parsing the string
+    '''
+
+    return datetime.strptime(s, '%Y-%m-%dT%H:%M:%SZ').replace(tzinfo=utc)
+
+
+class DataQuery(object):
+    r'''An object representing a query for data from a THREDDS server.
+
+    This object provides a clear API to formulate a query for data, including
+    a spatial query, a time query, and possibly some variables or other parameters.
+    These objects provide a dictionary-like interface, (``items`` and ``__iter``)
+    sufficient to be passed to functions expecting a dictionary representing a URL query.
+    Instances of this object can also be turned into a string, which will yield a
+    properly escaped string for a URL.
+    '''
+
+    def __init__(self):
+        r'Constructs an empty DataQuery'
+
+        self.var = set()
+        self.time_query = OrderedDict()
+        self.spatial_query = OrderedDict()
+        self.extra_params = OrderedDict()
+
+    def variables(self, *var_names):
+        r'''Specify one or more variables for the query.
+
+        This function ensures that variable names are not repeated.
+
+        This modifies the query in-place, but returns ``self`` so that multiple
+        queries can be chained together on one line.
+
+        Parameters
+        ----------
+        var_names : one or more strings
+            One or more names of variables to request
+
+        Returns
+        -------
+        self : ``DataQuery`` instance
+            Returns self for chaining calls
+        '''
+
+        self.var.update(set(var_names))
+        return self
+
+    def add_query_parameter(self, **kwargs):
+        r'''Add arbitrary query element (name=value) to the request.
+
+        This modifies the query in-place, but returns ``self`` so that multiple
+        queries can be chained together on one line.
+
+        Parameters
+        ----------
+        kwargs : one or more strings passed as keyword arguments
+            Names and values of parameters to add to the query
+
+        Returns
+        -------
+        self : ``DataQuery`` instance
+            Returns self for chaining calls
+        '''
+
+        self.extra_params.update(kwargs)
+        return self
+
+    def lonlat_box(self, north, south, east, west):
+        r'''Add a latitude/longitude bounding box to the query.
+
+        This adds a request for a spatial bounding box, bounded by (`south`, `north`)
+        for latitude and (`west`, `east`) for the longitude. This modifies the query
+        in-place, but returns ``self`` so that multiple queries can be chained together
+        on one line.
+
```

Comment:
We should probably mention the units on latitude and longitude - I think the TDS expects degree_east and degree_north.